gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaAvroSchemaRegistry.java

```diff
@@ -0,0 +1,289 @@
+    } else {
+      schemaId = headers[0].getValue().substring(SCHEMA_ID_HEADER_PREFIX.length());
+    }
+
+    if (response != null) {
+      LOG.info("Received response " + response);
+    }
+
+    return schemaId;
+  }
+
+  private class KafkaSchemaCacheLoader extends CacheLoader<String, Schema> {
+
+    private final Map<String, FailedFetchHistory> failedFetchHistories;
+
+    private KafkaSchemaCacheLoader() {
+      super();
+      this.failedFetchHistories = Maps.newHashMap();
+    }
+
+    @Override
+    public Schema load(String id) throws Exception {
+      if (shouldFetchFromSchemaRegistry(id)) {
+        try {
+          return fetchSchemaByID(id);
+        } catch (SchemaNotFoundException e) {
+          addFetchToFailureHistory(id);
+          throw e;
+        }
+      }
+      throw new SchemaNotFoundException(String.format("Schema with ID = %s cannot be retrieved", id));
+    }
+
+    private void addFetchToFailureHistory(String id) {
+      if (!failedFetchHistories.containsKey(id)) {
+        failedFetchHistories.put(id, new FailedFetchHistory(1, System.nanoTime()));
+      } else {
+        failedFetchHistories.get(id).incrementNumOfAttempts();
+        failedFetchHistories.get(id).setPreviousAttemptTime(System.nanoTime());
+      }
+    }
+
+    private boolean shouldFetchFromSchemaRegistry(String id) {
+      if (!failedFetchHistories.containsKey(id)) {
+        return true;
+      }
+      FailedFetchHistory failedFetchHistory = failedFetchHistories.get(id);
+      boolean maxTriesNotExceeded = failedFetchHistory.getNumOfAttempts() < GET_SCHEMA_BY_ID_MAX_TIRES;
+      boolean minRetryIntervalSatisfied =
+          System.nanoTime() - failedFetchHistory.getPreviousAttemptTime() >= TimeUnit.SECONDS
+              .toNanos(GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS);
+      return maxTriesNotExceeded && minRetryIntervalSatisfied;
+    }
+
+    private Schema fetchSchemaByID(String id) throws SchemaNotFoundException {
+      String schemaUrl = KafkaAvroSchemaRegistry.this.url + GET_RESOURCE_BY_ID + id;
+
+      GetMethod get = new GetMethod(schemaUrl);
+
+      int statusCode;
+      String schemaString;
+      try {
+        statusCode = KafkaAvroSchemaRegistry.this.httpClient.executeMethod(get);
+        schemaString = get.getResponseBodyAsString();
+      } catch (HttpException e) {
+        throw new RuntimeException(e);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      } finally {
+        get.releaseConnection();
+      }
+
+      if (statusCode != HttpStatus.SC_OK) {
+        throw new SchemaNotFoundException(String.format("Schema with ID = %s cannot be retrieved", id));
+      }
+
+      Schema schema;
+      if (schemaString.startsWith("{")) {
+        try {
+          schema = new Schema.Parser().parse(schemaString);
+        } catch (Exception e) {
+          throw new SchemaNotFoundException(String.format("Schema with ID = %s cannot be parsed", id));
```

Comment:
You might want to add an additional constructor to `SchemaNotFoundException` to take the exception it wraps as input.